crux_kv/src/lib.rs

```diff
@@ -2,27 +2,43 @@
 //!
 //! `crux_kv` allows Crux apps to store and retrieve arbitrary data by asking the Shell to
 //! persist the data using platform native capabilities (e.g. disk or web localStorage)
-//!
-//! This is still work in progress and extremely basic.
+
+pub mod error;
+
 use crux_core::capability::{CapabilityContext, Operation};
 use crux_core::macros::Capability;
+use error::KeyValueResult;
 use serde::{Deserialize, Serialize};
 
 /// Supported operations
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
 pub enum KeyValueOperation {
     /// Read bytes stored under a key
-    Read(String),
+    Get { key: String },
     /// Write bytes under a key
-    Write(String, Vec<u8>),
+    Set { key: String, value: Vec<u8> },
+    /// Remove a key and its value
+    Delete { key: String },
+    /// Test if a key exists
+    Exists { key: String },
 }
 
 #[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
 pub enum KeyValueOutput {
-    // TODO: Add support for errors
-    Read(Option<Vec<u8>>),
-    // TODO: Add support for errors
-    Write(bool),
+    Get {
+        value: KeyValueResult<Option<Vec<u8>>>,
+    },
+    Set {
+        result: KeyValueResult<()>,
+    },
+    Delete {
+        result: KeyValueResult<()>,
+    },
+    Exists {
+        result: KeyValueResult<bool>,
+    },
```

Comment:
Does the codegen handle that type correctly despite the generics...?

Reply:
Good catch. Thanks. No it doesn't! I replaced them with enums.